### ISSUE TITLE
Performance not improving when using CuDNN?

### ISSUE THREAD

**Issue by vedantroy** [NVIDIA/TransformerEngine#1113]
[Flash attention](https://github.com/Dao-AILab/flash-attention) shows CuDNN outperforms FlashAttention2.

But, I'm getting roughly similar performance with flash attention 2 and transformer engine with the CuDNN backend. Here's my benchmarking script:

```
from flash_attn.utils.benchmark import benchmark_forward, benchmark_backward
from flash_attn import (
    flash_attn_varlen_qkvpacked_func
)
from einops import rearrange
import numpy as np
import torch
import transformer_engine.pytorch as te

import os
# CudNN Graph api
os.environ["NVTE_FUSED_ATTN"] = "1"
os.environ["NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT"] = "1"


device = "cuda:0"
dtype = torch.bfloat16

B, N, L = 2, 10_000, 300
T, H, D = (B * (N + L)), 24, 64
# create 4 sequences, evenly sizedf
SEQ_SIZE = T // 4
seqlens = [SEQ_SIZE] * 3
seqlens.append(T - sum(seqlens))
assert np.sum(seqlens) == T, f"Sum of seqlens ({np.sum(seqlens)}) must equal T ({T})"

qkv = torch.randn(T, 3, H, D, device=device, dtype=dtype)
q, k, v = qkv.unbind(dim=1)

cu_seqlens_q_cum = np.zeros(len(seqlens) + 1, dtype=np.int32)
cu_seqlens_q_cum[1:] = np.cumsum(seqlens)
cu_seqlens_q_cum = torch.from_numpy(cu_seqlens_q_cum).to(device)

te_attn = te.DotProductAttention(
    num_attention_heads=H,
    kv_channels=D,
    qkv_format="thd", # tokens, head, dim
    attn_mask_type="padding",
)

te_out = te_attn(
    q, k, v,
    cu_seqlens_q=cu_seqlens_q_cum,
    max_seqlen_q=np.max(seqlens),
    max_seqlen_kv=np.max(seqlens),
    cu_seqlens_kv=cu_seqlens_q_cum
)

tols = dict(atol=2.5e-2, rtol=2.5e-2)

flash_out: torch.Tensor = flash_attn_varlen_qkvpacked_func(
    qkv,
    cu_seqlens=cu_seqlens_q_cum,
    max_seqlen=np.max(seqlens),
)

torch.testing.assert_close(flash_out, rearrange(te_out, 'T (H D) -> T H D', H=H), atol=5e-3, rtol=5e-3)

for i in range(2):
    print(f"bench {i}")
    _, flash_m = benchmark_forward(
        flash_attn_varlen_qkvpacked_func, qkv, cu_seqlens_q_cum, np.max(seqlens), repeats=50, verbose=False
    )
    
    _, te_m = benchmark_forward(
        te_attn, q, k, v, repeats=50, cu_seqlens_q=cu_seqlens_q_cum, max_seqlen_q=np.max(seqlens), max_seqlen_kv=np.max(seqlens), cu_seqlens_kv=cu_seqlens_q_cum, verbose=False
    )
    
    print(f"Flash: mean={flash_m.mean}, median={flash_m.median}")
    print(f"TE: mean={te_m.mean}, median={te_m.median}")


for i in range(2):
    print(f"bench bwd {i}")
    flash_grad = torch.randn_like(flash_out)
    te_grad = torch.randn_like(te_out)

    qkv.requires_grad_()
    q.requires_grad_()
    k.requires_grad_()
    v.requires_grad_()

    _, flash_m = benchmark_backward(
        flash_attn_varlen_qkvpacked_func, qkv, cu_seqlens_q_cum, np.max(seqlens), grad=flash_grad, repeats=50, verbose=False
    )
    
    _, te_m = benchmark_backward(
        te_attn, q, k, v, repeats=50, cu_seqlens_q=cu_seqlens_q_cum, max_seqlen_q=np.max(seqlens), max_seqlen_kv=np.max(seqlens), cu_seqlens_kv=cu_seqlens_q_cum, grad=te_grad, verbose=False
    )
    
    print(f"Flash: mean={flash_m.mean}, median={flash_m.median}")
    print(f"TE: mean={te_m.mean}, median={te_m.median}")
```

Outputs:
```
Flash: mean=0.0021125075197778644, median=0.0021125075197778644
TE: mean=0.002271024080691859, median=0.002271024080691859
bench 1
Flash: mean=0.002129286640556529, median=0.002129286640556529
TE: mean=0.002277279000263661, median=0.002277279000263661
bench bwd 0
Flash: mean=0.006021905019879341, median=0.006021905019879341
TE: mean=0.006118590539554134, median=0.006118590539554134
bench bwd 1
Flash: mean=0.006005273800110444, median=0.006005273800110444
TE: mean=0.0061185801401734356, median=0.0061185801401734356
```

Any thoughts on what I might be doing wrong?

**Edit:** I get a definite win at head dim 128. Maybe CuDNN just does better at larger head dims.
```
bench 0
Flash: mean=0.003518069520359859, median=0.003518069520359859
TE: mean=0.002689688680693507, median=0.002689688680693507
bench 1
Flash: mean=0.003591226399876177, median=0.003591226399876177
TE: mean=0.0026920175598934295, median=0.0026920175598934295
bench bwd 0
Flash: mean=0.010546553599415347, median=0.010546553599415347
TE: mean=0.007746712840162218, median=0.007746712840162218
bench bwd 1
Flash: mean=0.010531400180188938, median=0.010531400180188938
TE: mean=0.00786671765963547, median=0.00786671765963547
```

**Comment by vedantroy** [comment]
Re-opening this because I think performance of CuDNN should be exceeding flash attention 2 @ head dim 64, according to this graph:

![image](https://github.com/user-attachments/assets/bd62b20e-55a3-4506-a392-7a13e3e53e99)


Maybe I'm doing something wrong?

**Comment by cyanguwa** [comment]
Hi @vedantroy , I'm not surprised that cuDNN performs better at `head_dim = 128` than 64. It matches with my benchmarks too.

I'm not aware of any benchmark scripts that directly produced that performance chart (maybe FA has published it somewhere). But it'd be good to check the cuDNN versions used in their scripts. Also, whether the `qkv_format` is `bshd` or `thd`, and add causal mask to your tests as well. I think this is all the possible differences I can see between your tests and FA's.

**Comment by cyanguwa** [comment]
This might be FA's script: https://github.com/Dao-AILab/flash-attention/blob/main/hopper/benchmark_attn.py
